File: NetworkIntrusionSpark/src/main/scala/TrainModel.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.ml.feature.{StringIndexer, VectorAssembler, StandardScaler}
import org.apache.spark.ml.classification.RandomForestClassifier
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
import org.apache.spark.ml.Pipeline

object TrainModel {
  def main(args: Array[String]): Unit = {
    println("Initialisation de SparkSession pour l'entraînement ML (Scala)...")
    val spark = SparkSession.builder()
      .appName("Network Intrusion ML - RandomForest (Scala)")
      .config("spark.driver.memory", "4g")
      .master("local[*]")
      .getOrCreate()
      
    import spark.implicits._

    val inputPath = "../unified_dataset.parquet"

    println(s"Chargement des données depuis $inputPath...")
    val df = spark.read.parquet(inputPath)

    println("Nettoyage des données (suppression des valeurs manquantes/infinies)...")
    val dfClean = df.na.drop()

    println("Échantillonnage à 10% pour un entraînement rapide...")
    var dfSample = dfClean.sample(withReplacement = false, fraction = 0.1, seed = 42)

    val featureCols = dfSample.columns.filter(_ != "Label")

    for (c <- featureCols) {
      dfSample = dfSample.withColumn(c, col(c).cast("double"))
    }
    
    dfSample = dfSample.na.fill(0.0)

    println("Remplacement des valeurs infinies et NaN par 0.0...")
    for (c <- featureCols) {
      dfSample = dfSample.withColumn(
        c, 
        when(isnan(col(c)) || col(c).isNull || col(c) === Double.PositiveInfinity || col(c) === Double.NegativeInfinity, 0.0)
          .otherwise(col(c))
      )
    }

    println("Configuration du Pipeline ML (VectorAssembler -> StringIndexer -> RandomForest)...")
    val assembler = new VectorAssembler()
      .setInputCols(featureCols)
      .setOutputCol("rawFeatures")
      .setHandleInvalid("skip")

    val scaler = new StandardScaler()
      .setInputCol("rawFeatures")
      .setOutputCol("features")
      .setWithStd(true)
      .setWithMean(false)

    val indexer = new StringIndexer()
      .setInputCol("Label")
      .setOutputCol("labelIndex")
      .setHandleInvalid("skip")

    val rf = new RandomForestClassifier()
      .setLabelCol("labelIndex")
      .setFeaturesCol("features")
      .setNumTrees(20)
      .setMaxDepth(5)
      .setSeed(42)

    val pipeline = new Pipeline().setStages(Array(assembler, scaler, indexer, rf))

    println("Séparation des données en Train (70%) et Test (30%)...")
    val Array(trainData, testData) = dfSample.randomSplit(Array(0.7, 0.3), seed = 42)

    println("Entraînement du modèle (cela peut prendre quelques minutes)...")
    val model = pipeline.fit(trainData)

    println("Génération des prédictions sur le jeu de test...")
    val predictions = model.transform(testData)

    val evaluatorAcc = new MulticlassClassificationEvaluator()
      .setLabelCol("labelIndex")
      .setPredictionCol("prediction")
      .setMetricName("accuracy")

    val evaluatorF1 = new MulticlassClassificationEvaluator()
      .setLabelCol("labelIndex")
      .setPredictionCol("prediction")
      .setMetricName("f1")

    val accuracy = evaluatorAcc.evaluate(predictions)
    val f1Score = evaluatorF1.evaluate(predictions)

    println("\n" + "=" * 50)
    println("RÉSULTATS DE L'ÉVALUATION (Scala)")
    println("=" * 50)
    println(f"Accuracy : $accuracy%.4f")
    println(f"F1-Score : $f1Score%.4f")

    println("\nAperçu des prédictions :")
    predictions.select("Label", "labelIndex", "prediction", "probability").show(10, truncate = false)

    spark.stop()
  }
}

File: train_model.py
import argparse
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, when, isnan
from pyspark.ml.feature import StringIndexer, VectorAssembler, StandardScaler
from pyspark.ml.classification import RandomForestClassifier
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml import Pipeline

def main():
    # Le Parquet unifié est lu depuis HDFS (produit par process_data.py)
    parser = argparse.ArgumentParser(description="Entraînement RandomForest sur les données CIC-IDS-2017")
    parser.add_argument("--input", default="hdfs:///user/projet/cleaned_dataset.parquet",
                        help="Chemin HDFS du Parquet nettoyé (produit par process_data.py)")
    args = parser.parse_args()

    print("Initialisation de SparkSession pour l'entraînement ML...")
    spark = SparkSession.builder \
        .appName("Network Intrusion ML - RandomForest") \
        .config("spark.driver.memory", "4g") \
        .getOrCreate()

    print(f"Chargement des données depuis {args.input}...")
    df = spark.read.parquet(args.input)
    
    print("Nettoyage des données (suppression des valeurs manquantes/infinies)...")
    df_clean = df.dropna()
    
    print("Échantillonnage à 10% pour un entraînement rapide (à modifier pour le modèle final)...")
    df_sample = df_clean.sample(withReplacement=False, fraction=0.1, seed=42)
    
    # 2. Préparation des Features
    # Toutes les colonnes sauf "Label" sont des features.
    feature_cols = [c for c in df_sample.columns if c != "Label"]
    
    # Certains types peuvent être "string", on doit s'assurer que les features sont numériques.
    # Pour simplifier, on va caster toutes les features en Double.
    for c in feature_cols:
        df_sample = df_sample.withColumn(c, col(c).cast("double"))
        
    df_sample = df_sample.fillna(0.0) # Sécurité pour les casts échoués
    
    # Nettoyage robuste des valeurs infinies et NaN qui font planter StandardScaler
    print("Remplacement des valeurs infinies et NaN par 0.0...")
    for c in feature_cols:
        df_sample = df_sample.withColumn(c, when(isnan(col(c)) | col(c).isNull() | (col(c) == float('inf')) | (col(c) == float('-inf')), 0.0).otherwise(col(c)))
        
    print("Configuration du Pipeline ML (VectorAssembler -> StringIndexer -> RandomForest)...")
    
    # Assembler les features en un vecteur
    assembler = VectorAssembler(inputCols=feature_cols, outputCol="rawFeatures", handleInvalid="skip")
    
    # Mettre à l'échelle (Scaler) - Recommandé bien que RF y soit moins sensible
    scaler = StandardScaler(inputCol="rawFeatures", outputCol="features", withStd=True, withMean=False)
    
    # Encoder le Label textuel en index numérique (0.0, 1.0, 2.0...)
    indexer = StringIndexer(inputCol="Label", outputCol="labelIndex", handleInvalid="skip")
    
    # Modèle Random Forest
    # Limité à 20 arbres pour l'ébauche pour que ça soit rapide
    rf = RandomForestClassifier(labelCol="labelIndex", featuresCol="features", numTrees=20, maxDepth=5, seed=42)
    
    pipeline = Pipeline(stages=[assembler, scaler, indexer, rf])
    
    # 3. Séparation Train/Test
    print("Séparation des données en Train (70%) et Test (30%)...")
    train_data, test_data = df_sample.randomSplit([0.7, 0.3], seed=42)
    
    # 4. Entraînement
    print("Entraînement du modèle (cela peut prendre quelques minutes)...")
    model = pipeline.fit(train_data)
    
    # 5. Prédictions et Évaluation
    print("Génération des prédictions sur le jeu de test...")
    predictions = model.transform(test_data)
    
    evaluator_acc = MulticlassClassificationEvaluator(labelCol="labelIndex", predictionCol="prediction", metricName="accuracy")
    evaluator_f1 = MulticlassClassificationEvaluator(labelCol="labelIndex", predictionCol="prediction", metricName="f1")
    
    accuracy = evaluator_acc.evaluate(predictions)
    f1_score = evaluator_f1.evaluate(predictions)
    
    print("\n" + "="*50)
    print("RÉSULTATS DE L'ÉVALUATION (Ébauche)")
    print("="*50)
    print(f"Accuracy : {accuracy:.4f}")
    print(f"F1-Score : {f1_score:.4f}")
    
    print("\nAperçu des prédictions :")
    predictions.select("Label", "labelIndex", "prediction", "probability").show(10, truncate=False)
    
    spark.stop()

if __name__ == "__main__":
    main()
